pop-api/examples/dao/src/lib.rs

```
@@ -0,0 +1,364 @@
+			description: Vec<u8>,
+		) -> Result<(), Error> {
+			let caller = self.env().caller();
+			let contract = self.env().account_id();
+			let current_block = self.env().block_number();
+			let proposal_id: u32 = self.proposals_created.saturating_add(1);
+			let vote_end =
+				current_block.checked_add(self.voting_period).ok_or(Error::ArithmeticOverflow)?;
+			if description.len() >= STRINGLIMIT.into() {
+				return Err(Error::StringLimitReached);
+			}
+			let proposal = Proposal {
+				description,
+				vote_start: current_block,
+				vote_end,
+				yes_votes: 0,
+				no_votes: 0,
+				executed: false,
+				beneficiary,
+				amount,
+				proposal_id,
+			};
+
+			self.proposals.insert(proposal_id, &proposal);
+
+			self.env()
+				.emit_event(Created { id: proposal_id, creator: caller, admin: contract });
+
+			Ok(())
+		}
+
+		/// Allows Dao's members to vote for a proposal
+		///
+		/// # Parameters
+		/// - `proposal_id` - Identifier of the proposal
+		/// - `approve` - Indicates whether the vote is in favor (true) or against (false) the
+		///   proposal.
+		#[ink(message)]
+		pub fn vote(&mut self, proposal_id: u32, approve: bool) -> Result<(), Error> {
+			let caller = self.env().caller();
+			let current_block = self.env().block_number();
+			let now = self.env().block_number();
+			// let props = self.proposals.clone();
+
+			let mut proposal = self.proposals.get(proposal_id).ok_or(Error::ProposalNotFound)?;
+
+			if current_block < proposal.vote_start || current_block > proposal.vote_end {
+				return Err(Error::VotingPeriodEnded);
+			}
+
+			let member = self.members.get(caller).ok_or(Error::NotAMember)?;
+
+			if member.last_vote >= proposal.vote_start {
+				return Err(Error::AlreadyVoted);
+			}
+
+			if approve {
+				proposal.yes_votes = proposal.yes_votes.saturating_add(member.voting_power);
+			} else {
+				proposal.no_votes = proposal.no_votes.saturating_add(member.voting_power);
+			}
+
+			debug_assert!(proposal.yes_votes > 0);
```

This is incorrect. Imagine if someone use this contract for production use, this line of code will panic instead of returning a meaningful error.

---

debug is not used in production build